twitter_scrooge/twitter_scrooge.bzl

```diff
@@ -0,0 +1,269 @@
+        cjars +
+        [remote_jars_file,
+         only_transitive_thrift_srcs_file,
+         immediate_thrift_srcs_file],
+    outputs = [ctx.outputs.srcjar],
+    arguments = [
+      only_transitive_thrift_srcs_file.path,
+      immediate_thrift_srcs_file.path,
+      ctx.outputs.srcjar.path,
+      remote_jars_file.path,
+    ],
+    progress_message = "creating scrooge files %s" % ctx.label,
+  )
+
+  jars = _collect_scalaattr(ctx.attr.deps) # should be _collect_scalaattr
+
+  scalaattr = struct(outputs = None,
+                     transitive_runtime_deps = jars.transitive_runtime_deps, #TODO are we missing any runtime deps?
+                     transitive_compile_exports = jars.transitive_compile_exports + set(transitive_cjars),
+                     transitive_runtime_exports = jars.transitive_runtime_exports,
+                     )
+
+  transitive_srcjars = collect_srcjars(ctx.attr.deps) + collect_extra_srcjars(ctx.attr.deps)
+
+  srcjarsattr = struct(
+    srcjar = ctx.outputs.srcjar,
+    transitive_srcjars = transitive_srcjars,
+  )
+
+  return struct(
+    scala = scalaattr,
+    srcjars=srcjarsattr,
+    extra_information=[struct(
+      srcjars=srcjarsattr,
+      scrooge_srcjar=struct(transitive_owned_srcs = transitive_owned_srcs + immediate_thrift_srcs),
+    )],
+  )
+
+def _collect_scalaattr(targets):
+  transitive_runtime_deps = set()
+  transitive_compile_exports = set()
+  transitive_runtime_exports = set()
+  for target in targets:
+    if hasattr(target, "scala"):
+      transitive_runtime_deps += target.scala.transitive_runtime_deps
+      transitive_compile_exports += target.scala.transitive_compile_exports
+      if hasattr(target.scala.outputs, "ijar"):
+        transitive_compile_exports += [target.scala.outputs.ijar]
+      transitive_runtime_exports += target.scala.transitive_runtime_exports
+
+  return struct(
+    transitive_runtime_deps = transitive_runtime_deps,
+    transitive_compile_exports = transitive_compile_exports,
+    transitive_runtime_exports = transitive_runtime_exports,
+  )
+
+scrooge_scala_srcjar = rule(
+    _gen_scrooge_srcjar_impl,
+    attrs={
+        "deps": attr.label_list(mandatory=True),
+        #TODO we should think more about how we want to deal
+        #     with these sorts of things... this basically
+        #     is saying that we have a jar with a bunch
+        #     of thrifts that we want to depend on. Seems like
+        #     that should be a concern of thrift_library? we have
+        #     it here through becuase we need to show that it is
+        #     "covered," as well as needing the thrifts to
+        #     do the code gen.
+        "remote_jars": attr.label_list(),
+        "_scrooge_core": attr.label(
+            default=Label("@scrooge_core//jar"),
+            allow_files=True),
+        "_libthrift": attr.label(
+            default=Label("@libthrift//jar"),
+            allow_files=True),
+        "_scrooge_generator": attr.label(
```

Comment:
not that we're using the scala_binary we can clean up these deps a lot...doing so